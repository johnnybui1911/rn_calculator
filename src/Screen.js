import React from 'react'
import { SafeAreaView, View, Text } from 'react-native'
import { connect } from 'react-redux'
import Button from './Button'
import { inputButtons, LARGE_FONT, formatDisplayOutput, MAX_SIZE_DISPLAY, FONT_THRESHOLD } from './constants'
import styles from './styles'

export const Screen = ({ rootReducer }) => {
  const { inputValue, outputValue } = rootReducer
  const displayValue = formatDisplayOutput(inputValue, outputValue)
  const displayLength = displayValue.length > MAX_SIZE_DISPLAY ? MAX_SIZE_DISPLAY + 1 : displayValue.length // check length of display value
  const fontSize = displayLength > FONT_THRESHOLD ? LARGE_FONT - (displayLength % FONT_THRESHOLD) * 10 : LARGE_FONT // decrease fontSize when length > threshold

  return (
    <SafeAreaView style={styles.rootContainer}>
      {/* Display Screen */}
      <View style={styles.displayContainer}>
        <Text style={[styles.output, { fontSize }]}>{displayValue}</Text>
      </View>
      <View style={styles.buttonContainer}>
        <View style={styles.numberButtonConatiner}>
          {inputButtons.numbers.map((row, index) => {
            return (
              // Numbers Buttons and "C", "." Buttons
              <View key={index} style={{ flex: 1, flexDirection: 'row' }}>
                {row.map(item => (
                  <Button key={item} item={item} />
                ))}
              </View>
            )
          })}
        </View>
        <View style={styles.operatorButtonConatiner}>
          <View style={{ flex: 1 }}>
            {inputButtons.operators.map(item => (
              // Operators Button  + - * /
              <Button key={item} item={item} isOperator />
            ))}
          </View>
        </View>
      </View>
    </SafeAreaView>
  )
}
const mapStateToProps = state => ({
  rootReducer: state
})
export default connect(mapStateToProps)(Screen)
