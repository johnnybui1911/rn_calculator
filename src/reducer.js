import { NUMBER_PRESS, OPERATOR_PRESS, EQUAL_PRESS, CLEAR_PRESS, DOT_PRESS, MAX_SIZE_DISPLAY, formatResult } from './constants'

export const initialState = {
  inputValue: '0',
  outputValue: '',
  selectedSymbol: null,
  hasDot: false
}
export const rootReducer = (state = initialState, action) => {
  switch (action.type) {
    case NUMBER_PRESS: {
      const { selectedSymbol, outputValue, hasDot } = state
      const prevInputValue = state.inputValue
      const lastIdSymbol = selectedSymbol ? prevInputValue.lastIndexOf(selectedSymbol) : -1
      const lastInput = prevInputValue.slice(lastIdSymbol + 1, prevInputValue.length)
      // input out of upper bound -> no changes
      if (lastInput.indexOf('e') < 0 && lastInput.replace(/\D/g, '').length >= MAX_SIZE_DISPLAY && !outputValue) {
        return state
      }
      // user press number input
      return {
        ...state,
        outputValue: '',
        inputValue:
          prevInputValue === '0' || (outputValue && !selectedSymbol && !hasDot) // previous input is 0 or input is the output from the last operation
            ? String(action.inputValue)
            : prevInputValue + action.inputValue
      }
    }

    case OPERATOR_PRESS: {
      const { selectedSymbol } = action
      let { inputValue } = state
      const prevSelectedSymbol = state.selectedSymbol
      // user press multiple different operators continuously -> use the last one
      if (prevSelectedSymbol !== selectedSymbol) {
        const condition = prevSelectedSymbol && inputValue.slice(-1) === prevSelectedSymbol
        inputValue = condition ? inputValue.slice(0, inputValue.length - 1) : inputValue
        return {
          ...state,
          hasDot: false,
          selectedSymbol,
          inputValue: inputValue + selectedSymbol
        }
      } else {
        // user press multiple similar operators in one long operation right after press input number (ex: 1+2+3+4)
        if (inputValue.slice(-1) !== selectedSymbol) {
          return {
            ...state,
            hasDot: false,
            selectedSymbol,
            inputValue: inputValue + selectedSymbol
          }
        }
        // user press multiple similar operators continuously  -> no changes
        return state
      }
    }

    case EQUAL_PRESS: {
      let { inputValue, selectedSymbol } = state
      // user press "=" before press operator "+,-,*,/" -> return the last input
      if (!selectedSymbol) {
        return { ...state, outputValue: inputValue, hasDot: false }
      }
      // user press "=" after press operator "+,-,*,/" -> calculate the result
      let outputValue = eval(inputValue.slice(-1) === selectedSymbol ? inputValue.slice(0, -1) : inputValue)
      outputValue = formatResult(outputValue)
      return {
        ...state,
        hasDot: false,
        selectedSymbol: null,
        inputValue: outputValue,
        outputValue
      }
    }

    case DOT_PRESS: {
      const { selectedSymbol, outputValue } = state
      //user press "." after press "="
      if (outputValue && !selectedSymbol) {
        return {
          ...state,
          hasDot: true,
          inputValue: '0.',
          outputValue: ''
        }
      }
      return state.hasDot //user press "." after already press "."
        ? state
        : {
            // user press "." first time
            ...state,
            hasDot: true,
            inputValue: state.inputValue.slice(-1) === '.' ? state.inputValue : state.inputValue + '.',
            outputValue: ''
          }
    }

    case CLEAR_PRESS: {
      return initialState
    }

    default:
      return state
  }
}

export default rootReducer
