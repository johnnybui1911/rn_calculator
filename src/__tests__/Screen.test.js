import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { initialState } from '../reducer'
import { Screen } from '../Screen'

const mockState = {
  inputValue: '999999999',
  outputValue: '999999999',
  selectedSymbol: null,
  hasDot: false
}

describe('Screen Rendering', () => {
  let component = shallow(<Screen rootReducer={initialState} />)
  it('should match to snapshot', () => {
    expect(toJson(component)).toMatchSnapshot('Default Font Snapshot')
  })

  component = shallow(<Screen rootReducer={mockState} />)
  it('should match to snapshot with mock state', () => {
    expect(toJson(component)).toMatchSnapshot('Small Font Snapshot')
  })
})
