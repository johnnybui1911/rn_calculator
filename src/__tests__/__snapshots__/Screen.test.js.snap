// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen Rendering should match to snapshot with mock state: Small Font Snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "justifyContent": "flex-end",
        "padding": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 88,
          },
          Object {
            "fontSize": 48,
          },
        ]
      }
    >
      999,999,999
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
          "paddingLeft": 6,
          "paddingVertical": 6,
        }
      }
    >
      <View
        key="0"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={7}
          key="7"
        />
        <Connect(Component)
          item={8}
          key="8"
        />
        <Connect(Component)
          item={9}
          key="9"
        />
      </View>
      <View
        key="1"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={4}
          key="4"
        />
        <Connect(Component)
          item={5}
          key="5"
        />
        <Connect(Component)
          item={6}
          key="6"
        />
      </View>
      <View
        key="2"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={1}
          key="1"
        />
        <Connect(Component)
          item={2}
          key="2"
        />
        <Connect(Component)
          item={3}
          key="3"
        />
      </View>
      <View
        key="3"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item="C"
          key="C"
        />
        <Connect(Component)
          item={0}
          key="0"
        />
        <Connect(Component)
          item="."
          key="."
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "padding": 6,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Connect(Component)
          isOperator={true}
          item="/"
          key="/"
        />
        <Connect(Component)
          isOperator={true}
          item="*"
          key="*"
        />
        <Connect(Component)
          isOperator={true}
          item="+"
          key="+"
        />
        <Connect(Component)
          isOperator={true}
          item="-"
          key="-"
        />
        <Connect(Component)
          isOperator={true}
          item="="
          key="="
        />
      </View>
    </View>
  </View>
</SafeAreaView>
`;

exports[`Screen Rendering should match to snapshot: Default Font Snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "justifyContent": "flex-end",
        "padding": 12,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 88,
          },
          Object {
            "fontSize": 48,
          },
        ]
      }
    >
      999,999,999
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
          "paddingLeft": 6,
          "paddingVertical": 6,
        }
      }
    >
      <View
        key="0"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={7}
          key="7"
        />
        <Connect(Component)
          item={8}
          key="8"
        />
        <Connect(Component)
          item={9}
          key="9"
        />
      </View>
      <View
        key="1"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={4}
          key="4"
        />
        <Connect(Component)
          item={5}
          key="5"
        />
        <Connect(Component)
          item={6}
          key="6"
        />
      </View>
      <View
        key="2"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item={1}
          key="1"
        />
        <Connect(Component)
          item={2}
          key="2"
        />
        <Connect(Component)
          item={3}
          key="3"
        />
      </View>
      <View
        key="3"
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Connect(Component)
          item="C"
          key="C"
        />
        <Connect(Component)
          item={0}
          key="0"
        />
        <Connect(Component)
          item="."
          key="."
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "padding": 6,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Connect(Component)
          isOperator={true}
          item="/"
          key="/"
        />
        <Connect(Component)
          isOperator={true}
          item="*"
          key="*"
        />
        <Connect(Component)
          isOperator={true}
          item="+"
          key="+"
        />
        <Connect(Component)
          isOperator={true}
          item="-"
          key="-"
        />
        <Connect(Component)
          isOperator={true}
          item="="
          key="="
        />
      </View>
    </View>
  </View>
</SafeAreaView>
`;
